A43-sdk-authorization.md

```diff
@@ -0,0 +1,721 @@
+// previous internalInterceptors will not be replaced.
+func (i *FileWatcherInterceptor) updateInternalInterceptors() {
+  // ...
+}
+
+func (i *FileWatcherInterceptor) UnaryInterceptor(
+	ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) 
+		(resp interface{}, err error) {
+  // ...
+}
+
+func (i *FileWatcherInterceptor) StreamInterceptor(
+	srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) 
+		(err error) {
+  // ...
+}
+
+// Close cleans up resources allocated by the interceptors.
+func (i *FileWatcherInterceptor) Close() {
+  i.cancel()
+}
+```
+
+2. Application enables SDK authorization in gRPC Server
+
+- Policy is statically initialized
+
+```Go
+creds := credentials.NewServerTLSFromFile(certFile, keyFile)
+i, err := authz.NewStatic(authzPolicy)
+serverOpts := []grpc.ServerOption{
+  grpc.Creds(creds),
+  grpc.ChainUnaryInterceptor(i.UnaryInterceptor),
+  grpc.ChainStreamInterceptor(i.StreamInterceptor),
+}
+s := grpc.NewServer(serverOpts...)
+```
+
+- Policy is reloaded dynamically from file path
+
+```Go
+creds := credentials.NewServerTLSFromFile(certFile, keyFile)
+i, err := authz.NewFileWatcher(authzPolicyFile, 1 * time.Hour)
+serverOpts := []grpc.ServerOption{
+  grpc.Creds(creds),
+  grpc.ChainUnaryInterceptor(i.UnaryInterceptor),
+  grpc.ChainStreamInterceptor(i.StreamInterceptor),
+}
+s := grpc.NewServer(serverOpts...)
+// In the end, free up the resources used for policy refresh.
+defer i.Close()
+```
+
+#### Java
+
+1. Authorization Server Interceptors
+
+```Java
+package io.grpc.authz;
+
+// Class of authorization server interceptor for static policy.
+public final class AuthorizationServerInterceptor implements ServerInterceptor {
+  // Constructor
+  private AuthorizationServerInterceptor(String authorizationPolicy) {
+    // Creates authorization engines. An IllegalArgumentException will be thrown
+    // if the policy file is invalid.
+  }
+
+  @Override
+  public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call, Metadata headers,
+      ServerCallHandler<ReqT, RespT> next) {
+    // ...
+  }
+
+  public static AuthorizationServerInterceptor create(String authorizationPolicy)
+      throws IllegalArgumentException {
+    return new AuthorizationServerInterceptor(authorizationPolicy);
+  }
+}
+
+// Class of authorization server interceptor for policy from file with refresh
+// capability.
+public final class FileAuthorizationServerInterceptor implements ServerInterceptor {
+  private volatile AuthorizationServerInterceptor internalAuthzServerInterceptor;
+  private final String policyFile;
+  private FileTime lastModifiedTime;
+
+  // Constructor
+  private FileAuthorizationServerInterceptor(File policyFile) {
+    // Read policy from policyFile and create an internalAuthzServerInterceptor. An
+    // IOException or IllegalArgumentException will be thrown if the policy file
```

Comment:
I think for this case we'll just have IOException. If the policy is invalid we'll just wrap the exception in an IOException. IllegalArgumentException would be more appropriate if the file does not exist (since the file name is the only argument here), but FileNotFoundException is generally used for that case and is an IOException.

Reply:
Sounds good. Updated the proposal.